File: ocaml/lib/ast.ml
type node =
  | Program of program
  | Expression of expression
  | Statement of statement

and expression =
  | Identifier of identifier
  | Integer of int
  | Boolean of bool
  | String of string
  | Prefix of
      { operator : Token.t
      ; right : expression
      }
  | Infix of
      { left : expression
      ; operator : Token.t
      ; right : expression
      }
  | If of
      { condition : expression
      ; consequence : block
      ; alternative : block option
      }
  | FunctionLiteral of
      { parameters : identifier list
      ; body : block
      }
  | Call of
      { fn : expression
      ; args : expression list
      }
  | Array of expression list
  | Index of
      { left : expression
      ; index : expression
      }
  | Hash of (expression * expression) list
  | Macro of
      { parameters : identifier list
      ; body : block
      }
[@@deriving show { with_path = false }, sexp]

and statement =
  | Let of
      { name : identifier
      ; value : expression
      }
  | Return of expression
  | ExpressionStatement of expression
  | BlockStatement of block
[@@deriving show { with_path = false }, sexp]

and identifier = { identifier : string } [@@deriving sexp]
and block = { block : statement list }
and program = { statements : statement list }

let token_literal = function
  | Program _ -> "program"
  | Expression _ -> "expression"
  | Statement _ -> "statement"
;;

File: haskell/src/AST.hs
module AST where

import Token

type Ast = [Token] -> Program             -- type synonym. 

newtype Program = Program [Statement]     -- _newtype_ keyword is similar to _type_ but with better compile time guarantees
    deriving (Show, Eq)                   -- It is used for structs with only one field.

data Statement                            -- enum Statement {
    = LetStatement Identifier Expression  --  LetStament(Identifier, Expression)
    | ReturnStatement Expression          --  ReturnStatement(Expression)
    | ExpressionStatement Expression      --  ...
    | BlockStatement Block
    | IllegalStatement
    deriving (Show, Eq)

data Expression
    = IdentExpr Identifier
    | IntLiteral Integer
    | BoolLiteral Bool
    | NegExpr Expression
    | NotExpr Expression
    | AddExpr Expression Expression
    | SubExpr Expression Expression
    | MulExpr Expression Expression
    | DivExpr Expression Expression
    | EqExpr Expression Expression
    | NeExpr Expression Expression
    | LtExpr Expression Expression
    | GtExpr Expression Expression
    | IfExpr Expression Block (Maybe Block)
    | FunctionExpr [Identifier] Block
    | CallExpr Expression [Expression]
    deriving (Show, Eq)

newtype Identifier = Identifier String
    deriving (Show, Eq)

newtype Block = Block [Statement]
    deriving (Show, Eq)
